### PR TITLE
feat(datasets): Added the Experimental SafetensorsDataset

### PR DESCRIPTION
**Pull request by MinuraPunchihewa**
## Description
This PR adds the `SafetensorsDataset` to support interactions with tensors stored in files in the Safetensors format.

Fixes https://github.com/kedro-org/kedro-plugins/issues/221

## Development notes
I have used the PickleDataset as a base for the implementation of this dataset (as it goes about the same manner to access files).

These changes have been tested,

1. Manually, by running the code locally to load and save tensors from and to Safetensors files.
2. Via the existing and newly added unit tests.

## Checklist

- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [X] Added tests to cover my changes
- [ ] Received approvals from at least half of the TSC (required for adding a new, non-experimental dataset)

### CODE REVIEW

**Comment by MinuraPunchihewa** [comment]
Hey @astrojuanlu,
This is my implementation of the `SafetensorsDataset`. I would appreciate a review.

**Comment by MinuraPunchihewa** [comment]
> Thank you for the PR, @MinuraPunchihewa! It looks great to me. I tested it manually, and everything works perfectly!

Thank you, @DimedS. I appreciate it. I just updated the release notes as well.

File: kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        >>> import torch
+        >>>
+        >>> data = {"embeddings": torch.zeros((10, 100)}
+        >>> dataset = SafetensorsDataset(
+        ...     filepath="test.safetensors",
+        ...     backend="torch"
+        ... )
+        >>> dataset.save(data)
+        >>> reloaded = dataset.load()
+        >>> assert data.equals(reloaded)
```
Comment (noklam) [inline kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py:51]:
```suggestion
        >>> from kedro_datasets_experimental.safetensors import SafetensorsDataset
        >>> import torch
        >>>
        >>> data = {"embeddings": torch.zeros((10, 100)}
        >>> dataset = SafetensorsDataset(
        ...     filepath="test.safetensors",
        ...     backend="torch"
        ... )
        >>> dataset.save(data)
        >>> reloaded = dataset.load()
        >>> assert data.equals(reloaded)
```

`assert data.equals(reloaded)` this gives me an error:
> AttributeError: 'dict' object has no attribute 'equals'

File: kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+    DEFAULT_LOAD_ARGS: dict[str, Any] = {}
+    DEFAULT_SAVE_ARGS: dict[str, Any] = {}
+    DEFAULT_FS_ARGS: dict[str, Any] = {"open_args_save": {"mode": "wb"}}
+
+    def __init__(  # noqa: PLR0913
+        self,
+        *,
+        filepath: str,
+        backend: str = "torch",
```
Comment (astrojuanlu) [inline kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py:62]: PyTorch is notably tricky to install, should we default to `numpy`? It's also the first backend shown in the docs https://pypi.org/project/safetensors/

Comment (astrojuanlu) [reply kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py:62]: And if we're choosing a default backend, should we add that to the dependencies? I think it's not too high stakes to depend on NumPy for this dataset.

Comment (astrojuanlu) [reply kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py:62]: Even the PyTorch backend seems to depend on NumPy

```
In [2]: ds = SafetensorsDataset(filepath="/tmp/output.safetensors", backend="torch")
/Users/juan_cano/.cache/uv/archive-v0/SRNuoVa2WtZC4Q79225OL/lib/python3.12/site-packages/torch/_subclasses/functional_tensor.py:295: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/utils/tensor_numpy.cpp:84.)
  cpu = _conversion_method_template(device=torch.device("cpu"))
```

**Review by astrojuanlu — APPROVED** [review]
Apart from the comments I made on dependencies and defaults, I tested this with local and S3 URLs, saving and loading data, loading data saved with one backend from another backend. It all seems to work. Thanks @MinuraPunchihewa !

File: kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py (hunk 1 above)
Comment (MinuraPunchihewa) [reply kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py]: Hey @noklam,
I made some updates to the example in the docstring. The above error should be fixed now.

File: kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py (hunk 2 above)
Comment (MinuraPunchihewa) [reply kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py]: @astrojuanlu That makes sense. I've updated the default backend to be `numpy` and added it to the dependencies. 
Do you want me to update the examples in the docstring to be `numpy` as well or should we leave them in `torch`?

Comment (astrojuanlu) [reply kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py]: Better to update the docstrings yes! @MinuraPunchihewa

Comment (MinuraPunchihewa) [reply kedro-datasets/kedro_datasets_experimental/safetensors/safetensors_dataset.py]: @astrojuanlu It's done!

**Comment by astrojuanlu** [comment]
This is going in, thanks @MinuraPunchihewa again!

**Pull request merged**